daml/Marketplace/Custody/Model.daml

```diff
@@ -120,23 +121,41 @@ template LifecycleRequest
           when (postAssetVersion /= assetDeposit.asset.id.version) $
             createNewVersionOfDescription description postAssetVersion serializedRemaining
           let payoutSplitQuantities = (\(qty, _) -> qty * assetDeposit.asset.quantity) <$> pending
-          -- Next step could be:
-          -- create SettlementInstruction with buyer = customer; seller = issuer; ..
           splitResultDeposits <- exercise safekeepingDepositCid AssetDeposit_Split with quantities = payoutSplitQuantities
           let payoutCids = init splitResultDeposits
-          payoutCids <- forA payoutCids \payoutDepositCid -> do
-            exerciseByKey @AssetSettlementRule assetDeposit.account.id $
-               AssetSettlement_AddController description.issuer
-            cid <- exerciseByKey @AssetSettlementRule description.safekeepingAccountId (
-              AssetSettlement_Transfer with
-                receiverAccountId = assetDeposit.account.id
-                depositCid = payoutDepositCid)
-            exerciseByKey @AssetSettlementRule assetDeposit.account.id $
-              AssetSettlement_RemoveController description.issuer
-            pure cid
+              masterAgreementSignatories = [provider, customer, investor]
+          payoutInstructions <- createPayoutInstructions payoutCids description assetDeposit operator masterAgreementSignatories
           assetDepositCid' <- create assetDeposit with asset = (assetDeposit.asset with id = (assetDeposit.asset.id with version = postAssetVersion))
           archive assetDepositCid
-          return $ assetDepositCid' :: payoutCids
+          return (assetDepositCid', payoutInstructions)
+
+createPayoutInstructions : [ContractId AssetDeposit] -> AssetDescription -> AssetDeposit -> Party -> [Party] -> Update [ContractId SettlementInstruction]
+createPayoutInstructions payoutCids assetDescription assetDeposit operator masterAgreementSignatories =
+  forA indexedPayoutCids
+    \(index, payoutDepositCid) ->
+      create SettlementInstruction with
+        asset = assetDeposit.asset
+        steps = [createSettlementDetails payoutDepositCid]
+        masterAgreement = masterAgreement
+        tradeId = createTradeId assetDeposit.asset.id index
+        observers = fromList $ operator :: masterAgreementSignatories
+        ..
+  where
+    indexedPayoutCids = zip [1..length payoutCids] payoutCids
```

Comment:
Use a `fold*` instead of `zip` + `forA`.

Reply:
You mean this?
```
createPayoutInstructions : [ContractId AssetDeposit] -> AssetDescription -> AssetDeposit -> Party -> [Party] -> Update [ContractId SettlementInstruction]
createPayoutInstructions payoutCids assetDescription assetDeposit operator masterAgreementSignatories =
    snd <$> foldlA createSI (1, []) payoutCids
  where
    masterAgreement = createMasterAgreement assetDescription assetDeposit masterAgreementSignatories
    createSettlementDetails payoutDepositCid =
      SettlementDetails with
        senderAccount = assetDescription.safekeepingAccount
        receiverAccount = assetDeposit.account
        depositCid = Some payoutDepositCid
    createSI (index, accumulator) payoutDepositCid = do
      siCid <- create SettlementInstruction with
        asset = assetDeposit.asset
        steps = [createSettlementDetails payoutDepositCid]
        masterAgreement = masterAgreement
        tradeId = createTradeId assetDeposit.asset.id index
        observers = fromList $ operator :: masterAgreementSignatories
        ..
      pure (index + 1, siCid :: accumulator)
      ```